### PR TITLE
8268151: Vector API toShuffle optimization

### PR DESCRIPTION
**Pull request by sviswa7**
The Vector API toShuffle method can be optimized  using existing vector conversion intrinsic.

The following changes are made:
1) vector.toShuffle java implementation is changed to call VectorSupport.convert.
2) The conversion intrinsic (inline_vector_convert()) in vectorIntrinsics.cpp is changed to allow shuffle as a destination type.
3) The shuffle.toVector intrinsic (inline_vector_shuffle_to_vector()) in vectorIntrinsics.cpp now explicitly generates conversion node instead of performing conversion during unbox. This is to remove unnecessary boxing during back to back vector.toShuffle and shuffle.toVector calls. 

Best Regards,
Sandhya

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8268151](https://bugs.openjdk.java.net/browse/JDK-8268151): Vector API toShuffle optimization

### Reviewers
 * [Paul Sandoz](https://openjdk.java.net/census#psandoz) (@PaulSandoz - **Reviewer**)
 * [Vladimir Ivanov](https://openjdk.java.net/census#vlivanov) (@iwanowww - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4326/head:pull/4326` \
`$ git checkout pull/4326`

Update a local copy of the PR: \
`$ git checkout pull/4326` \
`$ git pull https://git.openjdk.java.net/jdk pull/4326/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4326`

View PR using the GUI difftool: \
`$ git pr show -t 4326`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4326.diff">https://git.openjdk.java.net/jdk/pull/4326.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/opto/vectornode.cpp [hunk 1]
```diff
@@ -1243,10 +1243,7 @@ Node* VectorUnboxNode::Ideal(PhaseGVN* phase, bool can_reshape) {
           value = phase->transform(VectorStoreMaskNode::make(*phase, value, in_vt->element_basic_type(), in_vt->length()));
           return new VectorLoadMaskNode(value, out_vt);
         } else if (is_vector_shuffle) {
-          if (is_shuffle_to_vector()) {
-            // VectorUnbox (VectorBox vshuffle) ==> VectorCastB2X vshuffle
-            return new VectorCastB2XNode(value, out_vt);
-          } else {
+          if (!is_shuffle_to_vector()) {
```
Comment (XiaohongGong) [inline src/hotspot/share/opto/vectornode.cpp:1246]: Hi @sviswa7 , thanks for this change! I'm just curious whether `is_shuffle_to_vector()` is still needed for `VectorUnboxNode` with this change? It seems this flag can be removed, doesn't it?

File: src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Byte128Vector.java [hunk 2]
```diff
@@ -330,9 +330,9 @@ public final long reduceLanesToLong(VectorOperators.Associative op,
         return (long) super.reduceLanesTemplate(op, m);  // specialized
     }
 
-    @Override
     @ForceInline
-    public VectorShuffle<Byte> toShuffle() {
+    private final
+    VectorShuffle<Byte> toShuffleTemplate(AbstractSpecies<Byte> dsp) {
```
Comment (XiaohongGong) [inline src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Byte128Vector.java:335]: Is it better to move this template method to the super class like other APIs?

File: src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Byte128Vector.java [hunk 3]
```diff
@@ -341,6 +341,15 @@ public final long reduceLanesToLong(VectorOperators.Associative op,
         return VectorShuffle.fromArray(VSPECIES, sa, 0);
     }
 
+    @ForceInline
+    public VectorShuffle<Byte> toShuffle() {
+        return VectorSupport.convert(VectorSupport.VECTOR_OP_CAST,
+                                     this.getClass(), ETYPE, VLENGTH,
+                                     Byte128Shuffle.class, byte.class, VLENGTH,
+                                     this, VSPECIES,
+                                     Byte128Vector::toShuffleTemplate);
```
Comment (XiaohongGong) [inline src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Byte128Vector.java:350]: ditto

File: src/hotspot/share/opto/vectornode.cpp (hunk 1 above)
Comment (sviswa7) [reply src/hotspot/share/opto/vectornode.cpp:1246]: @XiaohongGong is_shuffle_to_vector is still needed as we shouldn't generate VectorLoadShuffleNode for shuffle.toVector.

File: src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Byte128Vector.java (hunk 2 above)
Comment (sviswa7) [reply src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Byte128Vector.java]: Yes, can be moved to super class. Done in the updated commit.

File: src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Byte128Vector.java (hunk 3 above)
Comment (sviswa7) [reply src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Byte128Vector.java]: Yes, can be moved to super class. Done in the updated commit.

**Review by PaulSandoz — APPROVED** [review]
Java changes look good.

**Comment by sviswa7** [comment]
@PaulSandoz Thanks a lot for the review.

File: src/hotspot/share/opto/vectornode.cpp (hunk 1 above)
Comment (XiaohongGong) [reply src/hotspot/share/opto/vectornode.cpp:1246]: OK, got it. Thanks!

**Comment by XiaohongGong** [comment]
Looks good to me. Thanks!

**Review by iwanowww — APPROVED** [review]
Looks good. 

One inefficiency I noticed is that repeated `toVector()`/`toShuffle` leave a trail of redundant `VectorCastB2X`/`VectorCast[S..L]2X` nodes behind.

**Comment by sviswa7** [comment]
@iwanowww @XiaohongGong Thanks a lot for the review.
@iwanowww I will take up the redundant VectorCastB2X/VectorCast[S..L]2X conversion optimizations separately.

**Pull request closed without merging**